### PR TITLE
New Resource: Waypoint Action

### PR DESCRIPTION
**Pull request by HenryEstberg**
### :hammer_and_wrench: This adds a new resource `hcp_waypoint_action`.

Example config:
```
resource "hcp_waypoint_action" "test" {
  name = "some_test_action"
  description = "some description for fun"
  request = {
    custom = {
      method = "GET"
      url= "https://www.example.com"
      body = "some body"
      headers = {
        Content-Type = "application/json"
      }
    }
  }
}
```

Acceptance testing:

```sh
$ make testacc TESTARGS='-run=TestAccWaypoint_Action_basic'

...
```

### CODE REVIEW

File: internal/provider/waypoint/resource_waypoint_action_config.go
Comment (catsby) [inline internal/provider/waypoint/resource_waypoint_action_config.go:418]: I think we should always set these, to detect drift

File: internal/provider/waypoint/resource_waypoint_action_config.go
Comment (catsby) [inline internal/provider/waypoint/resource_waypoint_action_config.go:520]: We should always read these and send them for update, I think? I just tried this branch out and confirmed that if you create a config with a description, and then update the tf configuration to remove the description, Terraform reports that it will remove the value but after successfully applying, the value remains unchanged. Every `terraform plan` afterwards shows a perpetual diff, because we never send the updated `""` value to the API

File: internal/provider/waypoint/data_source_waypoint_action_config.go
Comment (HenryEstberg) [reply internal/provider/waypoint/data_source_waypoint_action_config.go:184]: The problem that I've encountered with this is that in some cases we do not set these, so the provider has them in state as `null` but the API sends back the empty string `""`. In this case, we really do want to ignore the empty string because if we don't, we get an error like this: `When applying changes to hcp_waypoint_action_config.test, provider
        "provider[\"registry.terraform.io/hashicorp/hcp\"]" produced an unexpected
        new value: .action_url: was null, but now cty.StringVal("")`

File: internal/provider/waypoint/data_source_waypoint_action_config.go
Comment (HenryEstberg) [reply internal/provider/waypoint/data_source_waypoint_action_config.go]: Yes, probably better to error out here than to let the process continue.

File: internal/provider/waypoint/data_source_waypoint_action_config.go
Comment (HenryEstberg) [reply internal/provider/waypoint/data_source_waypoint_action_config.go]: Not really, turns out we will not be adding those in the near future.

File: internal/provider/waypoint/resource_waypoint_action_config.go
Comment (HenryEstberg) [reply internal/provider/waypoint/resource_waypoint_action_config.go]: Good catch, I don't see these being used elsewhere.

File: internal/provider/waypoint/resource_waypoint_action_config.go
Comment (HenryEstberg) [reply internal/provider/waypoint/resource_waypoint_action_config.go:236]: This is actually necessary, when trying to use the assignment `:=` in those locations the compiler complains so without the initial declaration it will not work.

File: internal/provider/waypoint/resource_waypoint_action.go
Comment (briancain) [inline internal/provider/waypoint/resource_waypoint_action.go:129]: This could use a better description...I'm also not entirely sure where a user finds this value in HCP.

File: internal/provider/waypoint/resource_waypoint_action.go
Comment (briancain) [inline internal/provider/waypoint/resource_waypoint_action.go:136]: Part of me is thinking maybe we don't allow users to set this at all, does it need to be present in the schema because it exists under the Action config, or is it just a field that's user-set-able? If it's the second one, maybe we remove it from the provider.

File: internal/provider/waypoint/resource_waypoint_action.go
Comment (briancain) [inline internal/provider/waypoint/resource_waypoint_action.go:147]: `schema.StringAttribute` might have a way to say like "always upper-case this value" or something?

File: internal/provider/waypoint/resource_waypoint_action.go
Comment (briancain) [reply internal/provider/waypoint/resource_waypoint_action.go:129]: It also looks like we don't use the namespace_id field anyway for our namespace, but use Org and Project IDs to look it up. So maybe this field goes away?

File: internal/provider/waypoint/resource_waypoint_action.go
Comment (briancain) [inline internal/provider/waypoint/resource_waypoint_action.go:244]: Should we include the `err` too here?

File: internal/provider/waypoint/resource_waypoint_action.go
Comment (HenryEstberg) [reply internal/provider/waypoint/resource_waypoint_action.go:226]: I like that idea, but I think that change is outside the scope of this PR. Currently the pattern for every Waypoint resource is to keep the two names separate, so I would not want to make that change here without changing the other resources. In addition, from what I have seen the current pattern of keeping the two seperate is also consistent across the rest of the `hcp` provider.

File: internal/provider/waypoint/resource_waypoint_action.go
Comment (HenryEstberg) [reply internal/provider/waypoint/resource_waypoint_action.go:129]: It was only used to expose the namespace_id to the user, I never intended it for our use. Your comment made me realize that the field is not that useful however so I will just remove it.

File: internal/provider/waypoint/resource_waypoint_action.go
Comment (HenryEstberg) [reply internal/provider/waypoint/resource_waypoint_action.go:136]: This could certainly be removed, it is present in the provider because the action config endpoint accepts it and I wasn't sure if it was redundant with the URL within the `request`. I will get rid of it in that case, the provider will function fine without it.

File: internal/provider/waypoint/resource_waypoint_action.go
Comment (briancain) [reply internal/provider/waypoint/resource_waypoint_action.go:226]: Got it, makes sense!

File: internal/provider/waypoint/resource_waypoint_action.go
Comment (briancain) [reply internal/provider/waypoint/resource_waypoint_action.go:136]: Yep, originally before the `Request` struct was present in actions, we just had `action_url`. Now it's kind of turned into a shortcut, where if you set it with no Request struct, we auto-create a simple GET request on that url. But maybe for the TF provider that shortcut might be more confusing 🤔

File: internal/provider/waypoint/resource_waypoint_action.go
Comment (HenryEstberg) [reply internal/provider/waypoint/resource_waypoint_action.go:147]: I found it! There is a built-in `Validators` field for every attribute!

File: internal/provider/waypoint/resource_waypoint_action.go
Comment (HenryEstberg) [reply internal/provider/waypoint/resource_waypoint_action.go:244]: The error that is returned is pretty redundant here, it is just a less verbose version of the error that is already added so I would say not in this case.

**Comment by HenryEstberg** [comment]
Hey @manish-hashicorp could we please get an approval on this one? Thanks!

**Review by briancain — APPROVED** [review]

File: internal/provider/waypoint/data_source_waypoint_action.go
Comment (catsby) [inline internal/provider/waypoint/data_source_waypoint_action.go:160]: I don't think we should do these checks here, instead we should just set the value to what the API gives us. For example (and hypothetical) if the `Description` were to be cleared out, this pattern would prevent the data source from updating it's value in state. As a rule we should just set the state to reflect the API

File: internal/provider/waypoint/data_source_waypoint_action.go
Comment (catsby) [inline internal/provider/waypoint/data_source_waypoint_action.go:185]: style nit: because the above `if er != nil` always ends in `return`, we can omit the `else` here. It makes the code easier to reason about _in general_, even if in this specific case the `else` branch is just 1 line

Comment (catsby) [reply internal/provider/waypoint/data_source_waypoint_action.go:185]: Specifically meaning this instead:

```golang
if err != nil {
  // stuff
  return
}
data.Request.Custom.Method = methodString
```

File: internal/provider/waypoint/data_source_waypoint_action.go
Comment (catsby) [inline internal/provider/waypoint/data_source_waypoint_action.go:205]: same comment here about always setting the value in state to whatever we get from the API

**Review by catsby — CHANGES_REQUESTED** [review]
I think we need to change the data source to always set the values

File: internal/provider/waypoint/resource_waypoint_action.go
Comment (catsby) [inline internal/provider/waypoint/resource_waypoint_action.go:140]: nit: If we're requiring/validating here, then the `convertMethodToStringType` and `convertMethodToEnumType` methods seem superfluous to me when we can convert them like so:
```golang
method := waypoint_models.HashicorpCloudWaypointActionConfigFlavorCustomMethod(plan.Request.Custom.Method.ValueString())
modelBody.ActionConfig.Request.Custom.Method = &method
```

later in the read we can just type cast I think

```golang
data.Request.Custom.Method = types.StringValue(string(*actionCfg.Request.Custom.Method))
```

File: internal/provider/waypoint/resource_waypoint_action.go
Comment (catsby) [inline internal/provider/waypoint/resource_waypoint_action.go:356]: same note on empty here.. if it gets edited remotely, the local state will not pick up on the change

**Review by catsby — COMMENTED** [review]
just ~2 more comments

**Comment by manish-hashicorp** [comment]
Will approve after waypoint team approves.

File: internal/provider/waypoint/resource_waypoint_action_config.go
Comment (catsby) [reply internal/provider/waypoint/resource_waypoint_action_config.go]: This file has changed since I commented, it's now `resource_waypoint_action.go`, but the comment remains:

on line 244 and 301 you can use `:=` and remove the `var diag` declaration since there's no need to declare it in the larger scope.

Comment (catsby) [reply internal/provider/waypoint/resource_waypoint_action_config.go]: I apologize, I should have first asked if you could share the compiler warning, could you paste that here if it's still complaining? Maybe I'm missing something!

File: internal/provider/waypoint/data_source_waypoint_action.go
Comment (HenryEstberg) [reply internal/provider/waypoint/data_source_waypoint_action.go:26]: Good catch, thank you for linking the documentation as well!

File: internal/provider/waypoint/data_source_waypoint_action.go
Comment (HenryEstberg) [reply internal/provider/waypoint/data_source_waypoint_action.go:160]: Makes sense to me, I will update how we are handling description those fields here so they can potentially be empty.

**Review by catsby — APPROVED** [review]

**Comment by HenryEstberg** [comment]
The Waypoint team has approved now @manish-hashicorp could you please take another look? Thank you!

**Review by WenInCode — APPROVED** [review]

**Pull request merged**
